### PR TITLE
[None][refactor] Refactor Kimi K3 MLP

### PR DESCRIPTION
**Pull request by jiaganc**
## Dev Engineer Review

- Refactors Kimi K3 dense and shared-expert MLP paths to use `GatedMLP` with `SituAndMul`.
- Adds conditional tensor-parallel sharding and combined direct-MoE reductions.
- Preserves communication-backed reductions, FP8 weight loading, FP32 router bias handling, and latent materialization before RMSNorm.
- Updates TP-aware KDA and MLA reductions, indexed state pools, FP8 decode finalization, padded MLA head loading, and checkpoint slicing.
- Moves reusable SiTU activation code to `tensorrt_llm/_torch/modules/situ.py`.
- Moves reference Kimi K3 MLP and routing code into test utilities.
- Removes obsolete production Kimi K3 MLP, RMSNorm, and gate modules.
- Adds validation for unsupported routing configurations.
- Adds the Kimi K3 MLP test to the B200 pre-merge test list.
- Review areas include TP rank and reduction behavior, FP8 loading and decode paths, `_Fp8BlockScaleWeightReadLinear` API compatibility, normalization ordering, and checkpoint slicing.
- The first CI run failed for commit `d62cfe4`. The second run for commit `b7f2089` has no reported completion status.

## QA Engineer Review

Changed test functions include:

- Added `test_gated_mlp_supports_fused_situ`.
- Added `test_kimi_k3_moe_output_matches_reference_across_parallel_modes`.
- Added `test_kimi_k3_dense_layer_uses_gated_mlp`.
- Renamed `test_fused_routing_matches_eager` to `test_fused_routing_matches_eager_reference`.
- Removed `test_mutation_controls_disable_fused_routing`.
- Removed `test_ineligible_configs_disable_fused_routing`.
- Added `test_invalid_config_rejected_by_production_routing`.
- Added coverage for non-contiguous input rejection in `situ_and_mul`.
- Updated SiTU and MoE routing tests and reference utilities.

Coverage status:

- `test_kimi_k3_mlp.py` is listed in `tests/integration/test_lists/test-db/l0_b200.yml`.
- `test_kimi_k3_moe_gate.py`, `test_kimi_k3_situ_and_mul.py`, and `test_kimi_k3_situ_moe.py` are not listed in the reported CI or manual-QA test lists.

Verdict: needs follow-up.

## Description

Refactor the Kimi K3 dense and shared-expert MLP paths to reuse TRT-LLM's standard `GatedMLP` with the K3 SiTU activation instead of maintaining a separate production MLP implementation.

- Tensor-parallelize the first dense MLP and shared experts when attention data parallelism is disabled; keep them replicated when attention data parallelism assigns different tokens to each rank. Dense MLP TP uses a node-local divisor of the intermediate size when the full TP size does not divide it.
- For direct MoE TP, overlap the shared and routed branches, concatenate their partial outputs, and issue one all-reduce before splitting the shared hidden-width result from the routed latent result. Communication-backed MoE paths retain the independent shared-expert reduction.
- Preserve the optional Kimi K3 FP8 weight-read path through a linear-compatible wrapper, keep the router correction bias in FP32, and materialize the routed latent slice before FlashInfer RMSNorm.
- Move `SituAndMul` into the shared modules package, colocate production Kimi K3 helpers with the model implementation, and move the legacy/reference sparse-MoE implementation into the unit-test tree.

This removes duplicated production MLP code while preserving checkpoint loading, FP8 conversion, attention-DP behavior, and the required ordering around routed-expert normalization and projection.

## Test Coverage

- Pre-commit hooks passed on the complete final diff.
- `python3 -m py_compile` passed for the changed Python modules and tests.
- Clean B200 build passed for CUDA architecture `100-real`.
- Changed Kimi K3 unit tests: `64 passed` on B200.
- Added or updated behavioral coverage for fused SiTU `GatedMLP`, dense/shared reduction-mode selection, TP shard loading, FP32 router bias, and the attention-DP replicated paths.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 1]
```diff
@@ -890,24 +1054,21 @@ def _routed_output():
...
             )
-            # EP partial latent sums are completed by the wrapper's own
-            # reduction BEFORE the (nonlinear) latent norm.
+            if self._use_combined_all_reduce:
+                return y
+            # EP partial latent sums must be completed before the nonlinear
+            # latent norm. Communication-backed paths combine internally;
+            # direct TP paths use the wrapper's AllReduce explicitly.
+            if self.routed_experts.comm is None and moe_all_reduce is not None:
+                y = moe_all_reduce(y)
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Updated in 7bee3f6efb. The runtime still rejects direct MoE TP when ConfigurableMoE does not provide a real reduction operator. I removed the tests that asserted collective placement or executed a mocked TP reduction. The replacement test follows the existing AFMoE/MiniMax construction-level pattern: it constructs attention-DP, single-rank, direct-TP8, and communication-backed modes and verifies the shared GatedMLP TP size/rank, shard shapes, and reduce_output contract. Pre-commit passes.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 2]
```diff
@@ -117,6 +127,10 @@
 _K3_MOE_TP_ENV = "TLLM_K3_MOE_TP_SIZE"
 _K3_MOE_EP_ENV = "TLLM_K3_MOE_EP_SIZE"
 
+# Route the exact Kimi RMSNorm through FlashInfer's single-kernel path. Set
+# this to "0" to retain the eager reference's byte-exact rounding behavior.
+_FUSED_RMSNORM = os.environ.get("KIMI_K3_FUSED_RMSNORM", "1") == "1"
```
Comment (WeiHaocheng) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Do we still need this flag?

Comment (jiaganc) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py]: Thanks for reviewing! This flag is no longer needed. Removed it in 3cdbf64f64 together with the now-unused IS_FLASHINFER_AVAILABLE import and fused branch. Production RMSNorm calls already use the shared RMSNorm implementation, while KimiK3RMSNorm instances in the attention-residual path are consumed only through .weight/.eps. Its eager forward remains for the test-only reference path.

**Review by yingguo-trt — APPROVED** [review]

File: tests/unittest/_torch/modules/moe/test_kimi_k3_mlp.py [hunk 3]
```diff
@@ -115,3 +140,203 @@ def test_gate_up_half_swap_mutation_breaks_accuracy():
...
+    torch.testing.assert_close(gated(x), reference(x), rtol=1.6e-2, atol=1e-3)
+
+
+@pytest.mark.parametrize(
+    "attention_dp,tp_size,rank,has_routed_comm,expected_shared_tp,expected_shared_rank,expected_reduce",
+    [
+        (True, 8, 7, True, 1, 0, False),
+        (False, 1, 0, False, 1, 0, False),
+        (False, 8, 7, False, 8, 7, False),
+        (False, 8, 7, True, 8, 7, True),
```
Comment (sunnyqgg) [inline tests/unittest/_torch/modules/moe/test_kimi_k3_mlp.py]: The `routed_comm` case is unreachable: comm is only created under attention DP, so `use_shared_tp` with a non-None comm never occurs and the `reduce_output=True` branch is dead code. Remove this case and the dead branch

**Review by Barry-Delaney — APPROVED** [review]

**Review by lori-ren — APPROVED** [review]

File: tests/unittest/_torch/modules/moe/test_kimi_k3_situ_moe.py [hunk 4]
```diff
@@ -180,12 +179,19 @@ def test_kimi_gate_reuses_deepseek_v3_routing():
         gate.e_score_correction_bias.normal_(std=0.05)
     hidden_states = torch.randn(2, 7, config.hidden_size)
 
-    expected_ids, expected_weights = gate(hidden_states)
+    logits = gate.compute_logits(hidden_states)
+    scores = logits.sigmoid()
+    scores_for_choice = scores + gate.e_score_correction_bias.unsqueeze(0)
```
Comment (sunnyqgg) [inline tests/unittest/_torch/modules/moe/test_kimi_k3_situ_moe.py]: These inlined expected values duplicate the eager routing math in KimiK3ReferenceMoEGate. Remove the duplication and reuse that class.

File: tests/unittest/_torch/modules/moe/test_kimi_k3_moe_gate.py [hunk 5]
```diff
@@ -98,30 +74,12 @@ def test_fused_routing_matches_eager():
...
-        _GateConfig(num_experts_per_token=1),
+        (_GateConfig(moe_router_activation_func="softmax"), "requires sigmoid"),
+        (_GateConfig(moe_renormalize=False), "requires top-k weight renormalization"),
     ],
 )
-def test_ineligible_configs_disable_fused_routing(cfg):
-    # softmax scoring, grouped routing, renormalize off, and top_k == 1 all
-    # diverge from the fused kernel's fixed contract -> eager path.
-    assert KimiK3MoEGate(cfg)._use_fused_routing is False
+def test_invalid_config_rejected_by_production_routing(cfg, error):
```
Comment (sunnyqgg) [inline tests/unittest/_torch/modules/moe/test_kimi_k3_moe_gate.py]: This test only asserts trivial guard clauses and adds little value; we could remove it

File: tests/unittest/_torch/modules/moe/test_kimi_k3_situ_and_mul.py [hunk 6]
```diff
@@ -66,6 +67,16 @@ def test_situ_and_mul_strided_rows():
...
+    full = torch.randn(8, 1024, dtype=torch.bfloat16, device="cuda")
+    x = full[:, ::2]
+
+    with pytest.raises(ValueError, match="contiguous last dimension"):
+        torch.ops.trtllm.situ_and_mul(x, 4.0, 25.0)
+
+
 @requires_cuda
 @pytest.mark.parametrize("situ_beta,situ_linear_beta", _BETAS, ids=_BETA_IDS)
 def test_kimi_k3_mlp_fused_activation_matches_eager(situ_beta, situ_linear_beta):
```
Comment (sunnyqgg) [inline tests/unittest/_torch/modules/moe/test_kimi_k3_situ_and_mul.py]: With KimiK3MLP now test-only, test_kimi_k3_mlp_fused_activation_matches_eager is redundant: op-level test_situ_and_mul and production-level.test_gated_mlp_supports_fused_situ already cover the same math. we could remove it.

**Review by sunnyqgg — APPROVED** [review]
LGTM

**Review by WeiHaocheng — APPROVED** [review]
LGTM in modeling parts

File: tests/unittest/_torch/modules/moe/test_kimi_k3_mlp.py (hunk 3 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/moe/test_kimi_k3_mlp.py]: Removed the unreachable `routed_comm` construction case and the dead shared-expert `reduce_output=True` branch. Shared TP now always uses the combined all-reduce, and the construction test covers only attention DP, single-rank, and direct TP. Focused result: 75 passed.

File: tests/unittest/_torch/modules/moe/test_kimi_k3_situ_moe.py (hunk 4 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/moe/test_kimi_k3_situ_moe.py]: Removed this test entirely because it exercised the internal non-fused routing switch and duplicated stronger production routing parity coverage in `test_fused_routing_matches_eager_reference`. Focused result: 75 passed.

File: tests/unittest/_torch/modules/moe/test_kimi_k3_moe_gate.py (hunk 5 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/moe/test_kimi_k3_moe_gate.py]: Removed the trivial production-routing guard-clause test. Focused result: 75 passed.

File: tests/unittest/_torch/modules/moe/test_kimi_k3_situ_and_mul.py (hunk 6 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/moe/test_kimi_k3_situ_and_mul.py]: Removed `test_kimi_k3_mlp_fused_activation_matches_eager` and its docstring claim. The SiTU op parity test and production `GatedMLP` parity test retain the behavior coverage. Focused result: 75 passed.
